source-twilio/source_twilio/streams.py

```diff
@@ -0,0 +1,683 @@
+            slice_ = {
+                self.lower_boundary_filter_field: current_start.format(self.time_filter_template),
+                self.upper_boundary_filter_field: current_end.format(self.time_filter_template),
+            }
+            yield slice_
+            current_start = current_end + self.slice_granularity
+
+    def stream_slices(
+        self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        for super_slice in super().stream_slices(sync_mode=sync_mode, cursor_field=cursor_field, stream_state=stream_state):
+            for dt_range in self.generate_date_ranges():
+                slice_ = copy.deepcopy(super_slice) if super_slice else {}
+                slice_.update(dt_range)
+                yield slice_
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        lower_bound = stream_slice and stream_slice.get(self.lower_boundary_filter_field)
+        upper_bound = stream_slice and stream_slice.get(self.upper_boundary_filter_field)
+        if lower_bound:
+            params[self.lower_boundary_filter_field] = lower_bound
+        if upper_bound:
+            params[self.upper_boundary_filter_field] = upper_bound
+        return params
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        unsorted_records = []
+        # When slicing dates to search UsageRecords streams
+        # Twilio API will simply return a (20)404 error when not finding any data
+        # in the given slice. read_records will simply raise and break the stream.
+        # To prevent this, one can remove the upper_bound limit and allow the stream to search since the start_date
+        # i choose to maintain the stream behaviour to allow for faster searches before 2020 and simply catch and pass raises.
+        try:
+            for record in super().read_records(sync_mode, cursor_field, stream_slice, stream_state):
+                record[self.cursor_field] = pendulum.parse(record[self.cursor_field], strict=False).to_iso8601_string()
+                unsorted_records.append(record)
+            sorted_records = sorted(unsorted_records, key=lambda x: x[self.cursor_field])
+            for record in sorted_records:
+                if record[self.cursor_field] >= self.state.get(self.cursor_field, self._start_date):
+                    self._cursor_value = record[self.cursor_field]
+                    yield record
+        except HTTPError:
```

Comment:
Should we be ignoring _all_ HTTP errors here, or just the `20404` code? I think it would be better to minimally ignore errors, except for the known case, so that other unexpected conditions would be loudly announced to us. So I'd prefer only passing on the `20404` error case rather than all HTTP errors in general unless there's a good reason not to do that.

Reply:
Fixed in 96cfc2e427ad787ad8a7f5f560963b3ae181dbbd

Just to give my thought process in that:
I've read throught their [error docs](https://www.twilio.com/docs/api/errors) ( for about the 29999 code errors, after that it gets really product specific ). i know airbyte-cdk will handle 429 and 5XX error responses, so the 4xx were still not handled and to the point where the incremental streams starts to run (where this check is made), all errors related to invalid authentication  or similar ( like 20008 and on ) should have already been called, so i thought passing all HTTPError exceptions would mostly be used on 4XX TCP/Data not found errors

Reply:
But i agree it would be good to raise all other errors and see if that is the case